File: packages_py/provider_api_getters/src/provider_api_getters/health_check/providers/postgres_health_check.py
#!/usr/bin/env python3
"""
PostgreSQL Health Check - Standalone debugging script

Run directly: python -m provider_api_getters.health_check.providers.postgres_health_check
Or from project root: python packages_py/provider_api_getters/src/provider_api_getters/health_check/providers/postgres_health_check.py

Uses:
- static_config for YAML configuration
- PostgresApiToken for connection config resolution
- db_connection_postgres for connection management
"""
import asyncio
import json
import sys
from pathlib import Path

# ============================================================
# Handle both direct execution and module import
# ============================================================
if __name__ == "__main__":
    # Add src directory to path for direct execution
    _src_dir = Path(__file__).parent.parent.parent.parent
    if str(_src_dir) not in sys.path:
        sys.path.insert(0, str(_src_dir))
    from provider_api_getters.api_token import PostgresApiToken
else:
    # Relative import when used as module
    from ...api_token import PostgresApiToken

try:
    from db_connection_postgres import (
        DatabaseConfig,
        init_db,
        close_db,
        get_health_status,
    )
except ImportError:
    print("[ERROR] db_connection_postgres package not found.")
    print("Ensure packages_py/db_connection_postgres is installed.")
    sys.exit(1)


async def check_postgres_health(config: dict = None) -> dict:
    """
    Check PostgreSQL connectivity using db_connection_postgres package.

    Args:
        config: Configuration dict (if None, loads from static_config)

    Returns:
        dict: Health check result with success status and data/error
    """
    print(f"\n{'='*20} Starting PostgreSQL Health Check {'='*20}\n")
    
    # Load config if not provided
    if config is None:
        from static_config import config as static_config
        config = static_config

    print("=" * 60)
    print("POSTGRESQL HEALTH CHECK")
    print("=" * 60)

    # Initialize provider from config
    provider = PostgresApiToken(config)
    api_key_result = provider.get_api_key()
    connection_config = provider.get_connection_config()

    # Debug output
    connection_host = connection_config.get('host', 'N/A')
    connection_port = connection_config.get('port', 'N/A')
    connection_db = connection_config.get('database', 'N/A')
    connection_user = connection_config.get('username', 'N/A')

    print(f"\n[Config]")
    print(f"  Host: {connection_host}")
    print(f"  Port: {connection_port}")
    print(f"  Database: {connection_db}")
    print(f"  Username: {connection_user}")
    print(f"  Has credentials: {api_key_result.has_credentials}")
    print(f"  Is placeholder: {api_key_result.is_placeholder}")

    if not api_key_result.has_credentials or api_key_result.is_placeholder:
        print("\n[ERROR] Missing or placeholder credentials")
        return {"success": False, "error": "Missing credentials"}

    # Map connection_config to DatabaseConfig
    host = connection_config.get("host", "localhost")
    port = int(connection_config.get("port", 5432))
    database = connection_config.get("database", "postgres")
    username = connection_config.get("username", "postgres")
    password = connection_config.get("password") or api_key_result.raw_api_key
    
    # Handle SSL mapping
    # connection_config['ssl'] might be bool or string
    raw_ssl = connection_config.get("ssl", False)
    ssl_mode = "disable"
    if isinstance(raw_ssl, str):
        ssl_mode = raw_ssl
    elif raw_ssl is True:
        ssl_mode = "require" # Default to require if explicitly True
    
    print(f"\n[Connecting via db_connection_postgres]")
    print(f"  SSL Mode: {ssl_mode}")

    db_config = DatabaseConfig(
        host=host,
        port=port,
        database=database,
        user=username,
        password=password,
        ssl_mode=ssl_mode,
        echo=True # Enable echo for debug script
    )

    try:
        # Initialize the global DB manager with our config
        await init_db(db_config)
        
        # Run health check
        health = await get_health_status()
        
        # Must close connections
        await close_db()

        if health.get("status") == "healthy":
            pg_version = health.get('pg_version')
            print(f"\n[Connection Established]")
            print(f"  Version: {pg_version}")
            
            return {
                "success": True,
                "message": "Connected to PostgreSQL",
                "data": {
                    "host": host,
                    "port": port,
                    "database": database,
                    "version": pg_version,
                },
            }
        else:
            error_msg = health.get('error', 'Unknown error')
            print(f"\n[Connection Failed]")
            print(f"  Error: {error_msg}")
            return {
                "success": False,
                "error": error_msg,
            }

    except Exception as e:
        print(f"\n[Exception]")
        print(f"  {type(e).__name__}: {e}")
        # Ensure fallback close
        try:
            await close_db()
        except:
            pass
            
        return {
            "success": False,
            "error": str(e),
        }


if __name__ == "__main__":
    # Load config when run directly as standalone script
    PROJECT_ROOT = Path(__file__).parent.parent.parent.parent.parent.parent.parent
    CONFIG_DIR = PROJECT_ROOT / "common" / "config"

    from static_config import load_yaml_config, config as static_config
    load_yaml_config(config_dir=str(CONFIG_DIR))

    print("\n")
    result = asyncio.run(check_postgres_health(static_config))
    print("\n" + "=" * 60)
    print("RESULT")
    print("=" * 60)
    print(json.dumps(result, indent=2, default=str))

File: packages_mjs/provider_api_getters/src/health_check/providers/postgres_health_check.mjs
#!/usr/bin/env node
/**
 * PostgreSQL Health Check - Standalone debugging script
 *
 * Run directly: node postgres_health_check.mjs
 *
 * Uses:
 * - static_config for YAML configuration
 * - PostgresApiToken for connection config resolution
 * - pg for native PostgreSQL connection
 */
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..', '..', '..', '..', '..', '..', '..');
const CONFIG_DIR = path.join(PROJECT_ROOT, 'common', 'config');

// ============================================================
// Load static config FIRST
// ============================================================
import { loadYamlConfig, config as staticConfig } from '@internal/app-static-config-yaml';
await loadYamlConfig({ configDir: CONFIG_DIR });

// ============================================================
// Provider API getter
// ============================================================
import { PostgresApiToken } from '../../api_token/index.mjs';

export async function checkPostgresHealth() {
  console.log('='.repeat(60));
  console.log('POSTGRESQL HEALTH CHECK');
  console.log('='.repeat(60));

  // Initialize provider from static config
  const provider = new PostgresApiToken(staticConfig);
  const apiKeyResult = provider.getApiKey();
  const connectionConfig = provider.getConnectionConfig();

  // Debug output
  console.log('\n[Config]');
  console.log(`  Host: ${connectionConfig.host || 'N/A'}`);
  console.log(`  Port: ${connectionConfig.port || 'N/A'}`);
  console.log(`  Database: ${connectionConfig.database || 'N/A'}`);
  console.log(`  Username: ${connectionConfig.username || 'N/A'}`);
  console.log(`  Has credentials: ${apiKeyResult.hasCredentials}`);
  console.log(`  Is placeholder: ${apiKeyResult.isPlaceholder}`);

  if (!apiKeyResult.hasCredentials || apiKeyResult.isPlaceholder) {
    console.log('\n[ERROR] Missing or placeholder credentials');
    return { success: false, error: 'Missing credentials' };
  }

  // Try to import pg
  let pg;
  try {
    pg = await import('pg');
  } catch {
    console.log('\n[ERROR] pg not installed');
    console.log('  Install with: npm install pg');
    return { success: false, error: 'pg not installed' };
  }

  const { Client } = pg.default || pg;

  // Build connection parameters
  const host = connectionConfig.host || 'localhost';
  const port = connectionConfig.port || 5432;
  const database = connectionConfig.database || 'postgres';
  const username = connectionConfig.username || 'postgres';
  const password = apiKeyResult.apiKey;

  // Check if SSL should be disabled via environment variables
  // SSL_CERT_VERIFY=0 or NODE_TLS_REJECT_UNAUTHORIZED=0 means disable SSL
  const sslCertVerify = process.env.SSL_CERT_VERIFY || '';
  const nodeTls = process.env.NODE_TLS_REJECT_UNAUTHORIZED || '';
  const disableSsl = sslCertVerify === '0' || nodeTls === '0';

  console.log('\n[Connecting]');
  console.log(`  postgresql://${username}:****@${host}:${port}/${database}`);
  console.log(`  SSL disabled: ${disableSsl} (SSL_CERT_VERIFY=${sslCertVerify || 'not set'}, NODE_TLS_REJECT_UNAUTHORIZED=${nodeTls || 'not set'})`);

  // Configure SSL: false = no SSL, undefined = let pg decide
  const sslConfig = disableSsl ? false : undefined;

  const client = new Client({
    host,
    port,
    database,
    user: username,
    password,
    connectionTimeoutMillis: 10000,
    ssl: sslConfig,
  });

  try {
    // Connect
    await client.connect();
    console.log('\n[Connection Established]');

    // Run health check query
    const result = await client.query('SELECT 1 as check');
    const versionResult = await client.query('SELECT version()');
    const version = versionResult.rows[0].version;

    console.log('\n[Query Results]');
    console.log(`  SELECT 1: ${result.rows[0].check}`);
    console.log(`  Version: ${version.length > 60 ? version.substring(0, 60) + '...' : version}`);

    return {
      success: true,
      message: 'Connected to PostgreSQL',
      data: {
        host,
        port,
        database,
        version,
      },
    };
  } catch (error) {
    if (error.code === '28P01') {
      console.log('\n[Authentication Error]');
      console.log(`  ${error.message}`);
      return {
        success: false,
        error: 'Invalid password',
      };
    } else if (error.code === '3D000') {
      console.log('\n[Database Error]');
      console.log(`  ${error.message}`);
      return {
        success: false,
        error: `Database '${database}' does not exist`,
      };
    } else if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
      console.log('\n[Connection Error]');
      console.log(`  ${error.message}`);
      return {
        success: false,
        error: `Cannot connect to ${host}:${port}`,
      };
    } else {
      console.log('\n[Exception]');
      console.log(`  ${error.name}: ${error.message}`);
      return {
        success: false,
        error: error.message,
      };
    }
  } finally {
    await client.end().catch(() => {});
  }
}

// Run if executed directly
if (process.argv[1] === __filename) {
  console.log('\n');
  const result = await checkPostgresHealth();
  console.log('\n' + '='.repeat(60));
  console.log('RESULT');
  console.log('='.repeat(60));
  console.log(JSON.stringify(result, null, 2));
}
